File: typescript_test_crate/tests/19_transition_forward.ts
class EventForwardTest {
    private _state: string;
    private _state_stack: Array<any>;
    private _state_context: Record<string, any>;
    private _return_value: any;
    private log: string[] =     [];

    constructor() {
        this._state_stack = [];
        this._state_context = {  };
        this._return_value = null;
        this.log =         [];
        this._state = "Idle";
        this._enter();
    }

    private _transition(target_state: string, exit_args: any = null, enter_args: any = null) {
        if (exit_args) {
            this._exit(...exit_args);
        } else {
            this._exit();
        }
        this._state = target_state;
        if (enter_args) {
            this._enter(...enter_args);
        } else {
            this._enter();
        }
    }

    private _change_state(target_state: string) {
        this._state = target_state;
    }

    private _dispatch_event(event: string, ...args: any[]) {
        const handler_name = `_s_${this._state}_${event}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            return handler.apply(this, args);
        }
    }

    private _enter(...args: any[]) {
        // No enter handlers
    }

    private _exit(...args: any[]) {
        // No exit handlers
    }

    public process() {
        this._dispatch_event("process");
    }

    public get_log(): string[] {
        this._return_value = null
        this._dispatch_event("get_log")
        return this._return_value
    }

    private _s_Idle_process() {
        this.log.push("idle:process:before");
        this._transition("Working", null, null);
        return this._dispatch_event("process");
        // This should NOT execute because -> => returns after dispatch
        this.log.push("idle:process:after");
    }

    private _s_Idle_get_log() {
        this._return_value = this.log;
        return this._return_value;;
    }

    private _s_Working_process() {
        this.log.push("working:process");
    }

    private _s_Working_get_log() {
        this._return_value = this.log;
        return this._return_value;;
    }
}


function main(): void {
    console.log("=== Test 19: Transition Forward (TypeScript) ===");
    const s = new EventForwardTest();
    s.process();
    const log = s.get_log();
    console.log(`Log: ${JSON.stringify(log)}`);

    // After transition-forward:
    // - Idle logs "idle:process:before"
    // - Transition to Working
    // - Working handles process(), logs "working:process"
    // - Return prevents "idle:process:after"
    if (!log.includes("idle:process:before")) {
        throw new Error(`Expected 'idle:process:before' in log: ${log}`);
    }
    if (!log.includes("working:process")) {
        throw new Error(`Expected 'working:process' in log: ${log}`);
    }
    if (log.includes("idle:process:after")) {
        throw new Error(`Should NOT have 'idle:process:after' in log: ${log}`);
    }

    console.log("PASS: Transition forward works correctly");
}

main();

File: rust_test_crate/tests/19_transition_forward.rs
use std::collections::HashMap;

pub struct EventForwardTest {
    _state: String,
    _state_stack: Vec<Box<dyn std::any::Any>>,
    _state_context: HashMap<String, Box<dyn std::any::Any>>,
    log: Vec<String>,
}

impl EventForwardTest {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _state_context: HashMap::from([]),
            log: Vec::new(),
            _state: String::from("Idle"),
        };
        this._enter();
        this
    }

    fn _transition(&mut self, target_state: &str) {
        self._exit();
        self._state = target_state.to_string();
        self._enter();
    }

    fn _dispatch_event(&mut self, event: &str) {
let handler_name = format!("_s_{}_{}", self._state, event);
// Rust requires match-based dispatch or a handler registry
// For now, use explicit match in caller;
    }

    fn _enter(&mut self) {
        // No enter handlers
    }

    fn _exit(&mut self) {
        // No exit handlers
    }

    pub fn process(&mut self) {
match self._state.as_str() {
            "Idle" => { self._s_Idle_process(); }
            "Working" => { self._s_Working_process(); }
            _ => {}
        }
    }

    pub fn get_log(&mut self) -> Vec<String> {
match self._state.as_str() {
            "Idle" => self._s_Idle_get_log(),
            "Working" => self._s_Working_get_log(),
            _ => Default::default(),
        }
    }

    fn _s_Idle_get_log(&mut self) -> Vec<String> {
return self.log.clone();
    }

    fn _s_Idle_process(&mut self) {
self.log.push("idle:process:before".to_string());
self._transition("Working");
return self._s_Working_process();
// This should NOT execute because -> => returns after dispatch
self.log.push("idle:process:after".to_string());
    }

    fn _s_Working_get_log(&mut self) -> Vec<String> {
return self.log.clone();
    }

    fn _s_Working_process(&mut self) {
self.log.push("working:process".to_string());
    }
}


fn main() {
    println!("=== Test 19: Transition Forward (Rust) ===");
    let mut s = EventForwardTest::new();
    s.process();
    let log = s.get_log();
    println!("Log: {:?}", log);

    // After transition-forward:
    // - Idle logs "idle:process:before"
    // - Transition to Working
    // - Working handles process(), logs "working:process"
    // - Return prevents "idle:process:after"
    assert!(log.contains(&"idle:process:before".to_string()), "Expected 'idle:process:before' in log: {:?}", log);
    assert!(log.contains(&"working:process".to_string()), "Expected 'working:process' in log: {:?}", log);
    assert!(!log.contains(&"idle:process:after".to_string()), "Should NOT have 'idle:process:after' in log: {:?}", log);

    println!("PASS: Transition forward works correctly");
}
